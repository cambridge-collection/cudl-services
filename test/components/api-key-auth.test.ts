import express from 'express';
import {
  getPassport,
  passportApiKeyAuthenticationComponent,
  passportInitialiseComponent,
  Users,
} from '../../src/components/api-key-auth';
import passport from 'passport';
import request from 'supertest';
import {StatusCodes} from 'http-status-codes';

describe('getPassport', () => {
  test('throws if no Authenticator is set', () => {
    const app = express();
    expect(() => getPassport(app)).toThrowErrorMatchingSnapshot();
  });

  test('returns registered Authenticator', () => {
    const authenticator = new passport.Authenticator();
    const app = express();
    app.set('passport', authenticator);
    expect(getPassport(app)).toBe(authenticator);
  });
});

describe('passportApiKeyAuthenticationComponent', () => {
  const users: Users = {
    secret123: {username: 'bob', email: 'bob@example.com'},
  };
  let app: express.Express;
  beforeEach(async () => {
    app = express();
    await passportInitialiseComponent.register(app);
  });

  test.each([
    [undefined, StatusCodes.UNAUTHORIZED],
    ['invalidkey', StatusCodes.UNAUTHORIZED],
    ['secret123', StatusCodes.OK],
  ])('causes requests to be authenticated', async (key, expectedStatus) => {
    await passportApiKeyAuthenticationComponent(users).register(app);

    app.use(getPassport(app).authenticate('token'), (req, res) => {
      res.end(`user: ${JSON.stringify(req.user)}`);
    });

    let req = request(app).get('/');
    if (key !== undefined) {
      req = req.set('x-token', key);
    }
    const resp = await req;
    expect(resp.status).toBe(expectedStatus);
    if (resp.ok) {
      expect(resp.text).toMatchInlineSnapshot(
        '"user: {\\"username\\":\\"bob\\",\\"email\\":\\"bob@example.com\\"}"'
      );
    }
  });

  describe('registration', () => {
    test('throws if passport is not registered', async () => {
      await expect(
        passportApiKeyAuthenticationComponent(users).register(express())
      ).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
