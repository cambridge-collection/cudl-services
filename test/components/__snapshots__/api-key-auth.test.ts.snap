// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPassport throws if no Authenticator is set 1`] = `"express app has no passport Authenticator on the \\"passport\\" setting. Has the passportInitialiseComponent been registered?"`;

exports[`passportApiKeyAuthenticationComponent registration throws if passport is not registered 1`] = `"express app has no passport Authenticator on the \\"passport\\" setting. Has the passportInitialiseComponent been registered?"`;
